sockeye/inference.py

```diff
@@ -1271,43 +1325,50 @@ def _beam_search(self,
                                                                        models_output_layer_w,
                                                                        models_output_layer_b)
 
-            # (2) compute length-normalized accumulated scores in place
-            if t == 1 and self.batch_size == 1:  # only one hypothesis at t==1
-                scores = scores[:1] / self.length_penalty(lengths[:1])
-            else:
-                # renormalize scores by length ...
-                scores = (scores + scores_accumulated * self.length_penalty(lengths - 1)) / self.length_penalty(lengths)
-                # ... but not for finished hyps.
-                # their predicted distribution is set to their accumulated scores at C.PAD_ID.
-                pad_dist[:, C.PAD_ID] = scores_accumulated[:, 0]
-                # this is equivalent to doing this in numpy:
-                #   pad_dist[finished, :] = np.inf
-                #   pad_dist[finished, C.PAD_ID] = scores_accumulated[finished]
-                scores = mx.nd.where(finished, pad_dist, scores)
-
-            # (3) get beam_size winning hypotheses for each sentence block separately
-            best_hyp_indices[:], best_word_indices[:], scores_accumulated[:, 0] = topk(scores, t=t)
+            # (2) Special treatment for finished and inactive rows. Inactive rows are inf everywhere;
+            # finished rows are inf everywhere except column zero, which holds the accumulated model score
+            scores += scores_accumulated
+            pad_dist[:, C.PAD_ID] = mx.nd.where(inactive, self.inf_array_long[:, 0], scores_accumulated[:, 0])
```

Comment:
this is the only place where we use `self.inf_array_long` and you need to get the 0th column here. Would it make sense to have `self.inf_array_long.shape` to be `(self.batch_size * self.beam_size,)` instead of `(self.batch_size * self.beam_size, 1)`?